scipy/stats/_hypotests.py

```diff
@@ -9,11 +9,101 @@
+
+            # if n_samples is None, all args are samples
+            n_samp = len(args) if n_samples is None else n_samples
+
+            # split samples from other arguments
+            samples = [np.atleast_1d(sample) for sample in args[:n_samp]]
+            args = args[n_samp:]
+
+            if axis is None:
+                samples = [sample.ravel() for sample in samples]
+                axis = 0
+
+            # if axis is not needed, just handle nan_policy and return
+            ndims = np.array([sample.ndim for sample in samples])
+            if np.all(ndims <= 1):
+                # Addresses nan_policy="propagate"
+                if nan_policy == 'propagate':
+                    return hypotest_fun_in(*samples, *args, **kwds)
+
+                # Addresses nan_policy="raise"
+                contains_nans = []
+                for sample in samples:
+                    contains_nan, _ = (
+                        scipy.stats.stats._contains_nan(sample, nan_policy))
+                    contains_nans.append(contains_nan)
+
+                # Addresses nan_policy="omit" (doesn't get here otherwise)
+                if any(contains_nans):
+                    # consider passing in contains_nans
+                    samples = _remove_nans(samples, paired)
+
+                return hypotest_fun_in(*samples, *args, **kwds)
+
+            # otherwise, concatenate all samples along axis, remembering where
+            # each separate sample begins
+            lengths = np.array([sample.shape[axis] for sample in samples])
+            split_indices = np.cumsum(lengths)
+            x = scipy.stats.stats._broadcast_concatenate(samples, axis)
+
+            # Addresses nan_policy="raise"
+            contains_nan, _ = (
+                scipy.stats.stats._contains_nan(x, nan_policy))
+
+            # Addresses nan_policy="omit"
+            # Assumes nan_policy="propagate" is handled by hypotest_fun_in
+            def hypotest_fun(x):
+                samples = np.split(x, split_indices)[:n_samp]
+                if nan_policy == 'omit' and contains_nan:
+                    samples = _remove_nans(samples, paired)
+                return hypotest_fun_in(*samples, *args, **kwds)
+
+            if np.all(ndims == 1) is None:
+                return hypotest_fun(x)
+            else:
+                x = np.moveaxis(x, axis, -1)  # needed by result_creator
+                res = np.apply_along_axis(hypotest_fun, axis=-1, arr=x)
+                return result_creator(res)
+
+        return vectorize_hypotest_wrapper
+    return vectorize_hypotest_decorator
+
 
 Epps_Singleton_2sampResult = namedtuple('Epps_Singleton_2sampResult',
                                         ('statistic', 'pvalue'))
 
 
+@_vectorize_hypotest_factory(
+    result_creator=lambda res: Epps_Singleton_2sampResult(res[..., 0],
+                                                          res[..., 1]),
```

Comment:
The `result_creator` seems redundant as it's almost all the time the same. Can you instead just pass the result class? If we would need to change the logic, we could have two arguments instead of one. `result_class` and `result_creator`. The creator doing a different logic for with the `result_class`.

Reply:
This is improved now. Now for most tests the wrapper factory just needs to take in the result class, e.g. `Epps_Singleton_2sampResult`. For tests that return a non-iterable, there also needs to be some sort of unpacking function.